### PR TITLE
Send ApplicationDownlinkSent upstream messages from TxAcks

### LINKED ISSUE #56: Send ApplicationDownlinkSent event from Tx acknowledgement
**Summary:** 

Send DownlinkSent event based on the Tx acknowledgement received from a Gateway Server.

**Why do we need this?**

To provide insight to the downstream to applications and integrations.

**What is already there? What do you see now?**

Tx acknowledgement is received by the GS.

**What is missing? What do you want to see?**

- The Tx acknowledgement should be sent from the GS to the NS
- The NS should figure out which application downlink belonged to that Tx acknowledgement, based on the correlation IDs. We probably need the GS to use (temporary) tokens to find the right correlation IDs

**How do you propose to implement this?**

- Add new rpc in `GsNs` for the Tx acknowledgement
- Send an event from GS to NS with the correlation IDs
- NS to send the DownlinkSent event to AS

**What can you do yourself and what do you need help with?**

@rvolosatovs if GS just gives you a correlation ID, you can't find the downlink message, right? I guess it's gone also anyway from the queue?

Should the GS hold the downlink message (in memory) and send the sent downlink message back to the NS?

How actually does the NS find the concerning downlink message in the case of a downlink NACK, because then it's also gone from the queue right?

----

Original issue: https://github.com/TheThingsIndustries/lorawan-stack/issues/1476 by @johanstokking

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #56, https://github.com/TheThingsNetwork/lorawan-stack/issues/3741

#### Changes

- Add `SessionKeyId` field to `DownlinkMessage` messages.
- NS stores "in-flight" downlink messages.
- When a TxAck is received from the GS, it tries to match with the original scheduled downlink. If found, an `ApplicationDownlinkSent` message is forwarded to the application.

#### Testing

Unit tests for downlink matching.

Test with actual device and verify that ApplicationDownlinkSent messages are received.

#### Notes for Reviewers

NS stores scheduled downlinks with a TTL of 60 seconds by default. I am not sure if this is too high or low.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/networkserver/redis/scheduled_downlink_matcher.go
Comment (johanstokking) [inline pkg/networkserver/redis/scheduled_downlink_matcher.go]: This variable is named too generically for this package. Also, godoc is invalid.

**Review by johanstokking — APPROVED** [review]
LGTM

Please wait for @adriansmares's approval too

File: pkg/networkserver/downlink.go
Comment (adriansmares) [inline pkg/networkserver/downlink.go]: This is correct, but I would have loved if we made the session key ID part of `genDown` (So add it to `generatedDownlink`).

File: pkg/networkserver/grpc_gsns.go
Comment (adriansmares) [inline pkg/networkserver/grpc_gsns.go:1289]: I think we will see quite some `not_found` errors on the GS side if we report the local `not_found` error if the TTL expires. Nonetheless, I think this is not terrible.

Consider making the timeout part of the NS config - at least we can configure this in production then.

File: pkg/networkserver/redis/scheduled_downlink_matcher.go
Comment (adriansmares) [inline pkg/networkserver/redis/scheduled_downlink_matcher.go:41]: This is the edgiest case, but what if someone intentionally adds `ns:downlink:mystaticthinghere` as a correlation ID ? Since the order of the correlation IDs is not defined, this may act weirder than it seems.

I think hashing the raw payload would be better, but I'm open to suggestions.

**Review by adriansmares — APPROVED** [review]
I think this works and is fine as is, but I have some small comments.

**Review by htdvisser — APPROVED** [review]
API addition LGTM.

**Review by adriansmares — APPROVED** [review]

**Pull request merged**
